firedrake/plot.py

```diff
@@ -873,3 +877,295 @@ def __call__(self, function):
+        elif degree == 2:
+            return np.array([[0, 3, 4, 1, 6, 5, 7, 2, 8],
+                             [9, 12, 13, 10, 15, 14, 16, 11, 17],
+                             [0, 9, 10, 1, 18, 11, 19, 2, 20],
+                             [3, 12, 13, 4, 21, 14, 22, 5, 23],
+                             [0, 9, 12, 3, 18, 15, 21, 6, 24],
+                             [1, 10, 13, 4, 19, 16, 22, 7, 25]]), "biquadratic"
+        else:
+            raise NotImplementedError(f"Not implemented for degree {degree} on {cell}")
+    else:
+        raise NotImplementedError(f"Not implemented for cell {cell}")
+
+
+def _pgfplot_create_patch_arrays(data, cell_node_list, cells, perms):
+    a = cell_node_list[cells]
+    offsets = np.tile((np.arange(a.shape[0]) * a.shape[1]).reshape(-1, 1), perms.shape[1])
+    return data[a.reshape(-1)[offsets + perms].reshape(-1), :]
+
+
+def _pgfplot_create_patches(f, coords, complex_component):
+    V = f.function_space()
+    elem = V.ufl_element()
+    fiat_cell = V.finat_element.cell
+    degree = elem.degree()
+    coordV = coords.function_space()
+    mesh = V.ufl_domain()
+    cdata = coords.dat.data_ro.real
+    fdata = f.dat.data_ro.real if complex_component == 'real' else f.dat.data_ro.imag
+    map_facet_dofs, patch_type = _pgfplot_make_perms(fiat_cell, degree)
+    if isinstance(fiat_cell, UFCTriangle):
+        cells = np.arange(mesh.cell_set.size)
+        perms = map_facet_dofs
+    elif isinstance(fiat_cell, UFCTetrahedron):
+        _facets = mesh.exterior_facets
+        nfacets = _facets.classes[1]
+        cells = _facets.facet_cell[:nfacets, 0]
+        perms = map_facet_dofs[_facets.local_facet_dat.data_ro[:nfacets]]
+    elif isinstance(fiat_cell, UFCQuadrilateral):
+        cells = np.arange(mesh.cell_set.size)
+        perms = map_facet_dofs
+    elif isinstance(fiat_cell, UFCHexahedron):
+        _facets = mesh.exterior_facets
+        nfacets = _facets.classes[1]
+        cells = _facets.facet_cell[:nfacets, 0]
+        perms = map_facet_dofs[_facets.local_facet_dat.data_ro[:nfacets]]
+    else:
+        raise NotImplementedError(f"Got unsupported FIAT cell: {fiat_cell}")
+    patches_c = _pgfplot_create_patch_arrays(cdata, coordV.cell_node_list, cells, perms)
+    patches_f = _pgfplot_create_patch_arrays(fdata.reshape(-1, 1), V.cell_node_list, cells, perms)
+    patches = np.concatenate([patches_c, patches_f], axis=1)
+    return patches, patch_type
+
+
+def pgfplot(f, filename, degree=1, complex_component='real', print_latex_example=True):
+    """Produce a data file for LaTeX tikz plotting in parallel.
+
+    Parameters
+    ----------
+    f
+       `Function` to plot.
+    filename
+        Name of the output file.
+    degree
+        Degree of interpolation for plotting: ``1`` (linear) or ``2`` (quadratic).
+    complex_component
+        Complex component to be plotted: 'real' or 'imag'.
+    print_latex_example
+        Flag indicating whether to print a latex example or not.
```

Comment:
Types on parameters please.

Reply:
Fixed! Also added one paragraph in the `Notes`.